### PR TITLE
Add aligned allocation and optimize checksumming of objects.

### PR DESCRIPTION
**Pull request by Goddesen**
This is effectively a rebase of #3955.

In short:
- Massive speedup of checksumming through autovectorization.
- Aligned and zero-padded memory allocation for object chunks.

### CODE REVIEW

**Comment by IntelOrca** [comment]
For MSVC, there is a `_aligned_malloc` which can be used instead.

**Comment by janisozaur** [comment]
C11 has it as part of standard: http://en.cppreference.com/w/c/memory/aligned_alloc

**Comment by IntelOrca** [comment]
If all our supported compilers work with aligned_alloc then there is no need to add a polyfill.

**Comment by Goddesen** [comment]
It will get preprocessor-y real quick, but I can whip up a commit and you can see what seems most practical.

File: src/core/Memory.hpp [hunk 1]
```diff
@@ -141,33 +141,26 @@ namespace Memory
...
+		void *ptr;
+#if defined(__GNUC__) || defined(__clang__)
+		int ret = posix_memalign(&ptr, alignment, size);
+		if (!ret) {
+#else
+		ptr = _aligned_malloc(size, alignment);
+		if (!ptr) {
+#endif
+			log_error("Failed to allocate aligned memory.");
+			assert(false);
```
Comment (Goddesen) [inline src/core/Memory.hpp]: This error handling could be better.

File: src/core/Memory.hpp [hunk 2]
```diff
@@ -141,33 +141,26 @@ namespace Memory
...
-		// Save alignment for realloc
-		((size_t *)((uint8 *)ptr_aligned - sizeof(uint8 *)))[-1] = alignment;
-		// Save original pointer for free
-		((uint8 **)ptr_aligned)[-1] = ptr;
-
-		// Zero-pad object up to a multiple of alignment
-		memset((uint8 *)ptr_aligned + size, 0x00, zero_pad);
+		void *ptr;
+#if defined(__GNUC__) || defined(__clang__)
+		int ret = posix_memalign(&ptr, alignment, size);
```
Comment (Goddesen) [inline src/core/Memory.hpp]: `aligned_alloc` is not available on Mac.

File: src/core/Memory.hpp [hunk 3]
```diff
@@ -137,4 +137,71 @@ namespace Memory
...
         Free(ptr);
     }
+
+	template<typename T>
+	T * AllocateAligned(size_t alignment, size_t size)
+	{
+		// All targets require alignment to be a power of two.
+		// Posix requires alignment to be at least as big as sizeof(void *).
+		// Return error on both for simplicity.
+		if (alignment % 2 || alignment < sizeof(void *)) {
```
Comment (IntelOrca) [inline src/core/Memory.hpp]: I'd use & instead of %.

File: src/core/Memory.hpp [hunk 4]
```diff
@@ -137,4 +137,71 @@ namespace Memory
...
+			return NULL;
```
Comment (IntelOrca) [inline src/core/Memory.hpp]: `nullptr`

File: src/core/Memory.hpp (hunk 1 above)
Comment (IntelOrca) [inline src/core/Memory.hpp]: Use `Console::` and `Guard::`

**Comment by Goddesen** [comment]
This should be ready for testing and/or more relentless scrutiny :)

**Comment by janisozaur** [comment]
I had hoped to find some time for testing it last weekend, but that didn't work out. I'll check at my earliest convenience.

File: src/core/Memory.hpp (hunk 3 above)
Comment (Goddesen) [inline src/core/Memory.hpp]: Derp. Will fix.

**Comment by janisozaur** [comment]
When rebased on current `develop` ab1708a, it fails one of our asserts when connecting to a server which has custom objects:

```
#0  0x00007ffff34a904f in raise () from /usr/lib/libc.so.6
No symbol table info available.
#1  0x00007ffff34aa47a in abort () from /usr/lib/libc.so.6
No symbol table info available.
#2  0x00007ffff34a1ea7 in __assert_fail_base () from /usr/lib/libc.so.6
No symbol table info available.
#3  0x00007ffff34a1f52 in __assert_fail () from /usr/lib/libc.so.6
No symbol table info available.
#4  0x0000000000f4521b in Guard::Assert_VA (expression=false, message=0x0, args=0x7fffffffcf70) at /home/janisozaur/workspace/OpenRCT2/src/core/Guard.cpp:92
        __PRETTY_FUNCTION__ = "void Guard::Assert_VA(bool, const char*, __va_list_tag*)"
#5  0x0000000000f45159 in Guard::Assert (expression=false, message=0x0) at /home/janisozaur/workspace/OpenRCT2/src/core/Guard.cpp:51
        args = <error reading variable args (Attempt to dereference a generic pointer.)>
#6  0x00000000010ad419 in Memory::AllocateAligned<void*> (alignment=32, size=6291488) at /home/janisozaur/workspace/OpenRCT2/src/object/../core/Memory.hpp:164
        ptr = 0x7fffcf21e800
        ret = 0
#7  0x00000000010a1f86 in allocate_chunk (size=6291456) at /home/janisozaur/workspace/OpenRCT2/src/object/ObjectRepository.cpp:999
        zero_pad = 32
        chunk = 0xffffffffa36
#8  0x000000000109ef36 in object_load_packed (rw=0x607000017630) at /home/janisozaur/workspace/OpenRCT2/src/object/ObjectRepository.cpp:762
        chunk = 0x7fffffffd640 "p\326\377\377\377\177"
        chunkSize = 140737488344064
        objRepo = 0x60c000018580
        entry = {
          flags = 1, 
          name = "A14WFC  ", 
          checksum = 1833764324
        }
        __func__ = "object_load_packed"
#9  0x000000000110c1d2 in S6Importer::LoadSavedGame (this=0x7fffc3a55800, rw=0x607000017630) at /home/janisozaur/workspace/OpenRCT2/src/rct2/S6Importer.cpp:119
        i = 0
#10 0x0000000001113cf1 in game_load_network (rw=0x607000017630) at /home/janisozaur/workspace/OpenRCT2/src/rct2/S6Importer.cpp:484
        result = false
        s6Importer = 0x7fffc3a55800
        checksum = 0
#11 0x0000000001048219 in Network::Client_Handle_MAP (this=0x2738780 <gNetwork>, connection=..., packet=...) at /home/janisozaur/workspace/OpenRCT2/src/network/network.cpp:1641
        has_to_free = true
        data = 0x7fffc178e800 "\003 "
        data_size = 7508016
        rw = 0x607000017630
        size = 1627096
        offset = 1625000
        chunksize = 2096
        str_downloading_map = "Downloading map ... (1588 / 1588) KiB\000\000\000R{\000\001\000\000\000\000<\b\000\000\000\000\000\000\000\b\024\000\020c", '\000' <repeats 18 times>, "\240\325\377\377\377\177\000\000\224n\000\001\000\000\000\000<\b\000\000\000\000\000\000\000\b\024\000\020c\000\000\260\325\377\377\377\177\000\000\374;\000\001\000\000\000\001\320\325\377\377\377\177\000\000 A\000\001\000\000\000\000\340\325\377\377\377\177\000\000`\337\000\000@`\000\000<\b\000\000\000\000\000\000\000\b\024\000\020c\000\000\060\326\377\377\377\177\000\000\223\366U\363\377\177\000\000<\b\000\000\000\000\000\000"...
        downloading_map_args = {[0] = 1588, [1] = 1588}
        __func__ = "Client_Handle_MAP"
#12 0x0000000001040e44 in Network::ProcessPacket (this=0x2738780 <gNetwork>, connection=..., packet=...) at /home/janisozaur/workspace/OpenRCT2/src/network/network.cpp:1212
        command = 0
#13 0x0000000001040634 in Network::ProcessConnection (this=0x2738780 <gNetwork>, connection=...) at /home/janisozaur/workspace/OpenRCT2/src/network/network.cpp:1174
        packetStatus = 0
#14 0x0000000001034ef2 in Network::UpdateClient (this=0x2738780 <gNetwork>) at /home/janisozaur/workspace/OpenRCT2/src/network/network.cpp:465
No locals.
#15 0x00000000010338b3 in Network::Update (this=0x2738780 <gNetwork>) at /home/janisozaur/workspace/OpenRCT2/src/network/network.cpp:369
No locals.
#16 0x000000000104bdfe in network_update () at /home/janisozaur/workspace/OpenRCT2/src/network/network.cpp:1985
No locals.
#17 0x00000000006a2f44 in rct2_update () at /home/janisozaur/workspace/OpenRCT2/src/rct2.c:459
        tickCount = 15086
#18 0x00000000005bb44e in openrct2_loop () at /home/janisozaur/workspace/OpenRCT2/src/openrct2.c:416
        nudge = 0.160000026
        is_minimised = false
        currentTick = 15086
        ticksElapsed = 32767
        lastTick = 0
        uncapTick = 15028
        fps = 7
        secondTick = 14822
#19 0x00000000005bb15e in openrct2_launch () at /home/janisozaur/workspace/OpenRCT2/src/openrct2.c:342
No locals.
#20 0x0000000000696414 in main (argc=1, argv=0x7fffffffdb78) at /home/janisozaur/workspace/OpenRCT2/src/platform/posix.c:57
        run_game = 1
```

I'll try taking a closer look.

**Comment by janisozaur** [comment]
`posix_memalign` returns 0 on success.

Upon fixing it:
```diff
@@ -155,7 +155,7 @@ namespace Memory
                if (!ptr) {
 #elif defined(__GNUC__) || defined(__clang__)
                int ret = posix_memalign(&ptr, alignment, size);
-               if (!ret) {
+               if (ret != 0) {
 #else
                ptr = _aligned_malloc(size, alignment);
                if (!ptr) {
```
there are still other errors:
```
ERROR[/home/janisozaur/workspace/OpenRCT2/src/localisation/localisation.c:1443 (win1252_to_utf8)]: Encountered invalid sequence
ERROR[/home/janisozaur/workspace/OpenRCT2/src/localisation/localisation.c:1443 (win1252_to_utf8)]: Encountered invalid sequence
[A14WFC  ] Error: Bad image table.
[A14WFC  ] Error: Unexpectedly reached end of file.
[A14WFC  ] Unable to export object.
[PATHWOOD] Error: Bad string table.
[PATHWOOD] Error: Unexpectedly reached end of file.
[PATHWOOD] Unable to export object.
<repeated numerous times>
ERROR[/home/janisozaur/workspace/OpenRCT2/src/localisation/localisation.c:1443 (win1252_to_utf8)]: Encountered invalid sequence
<repeated numerous times>
[AWATER  ] Error: Bad string table.
[AWATER  ] Error: Unexpectedly reached end of file.
[AWATER  ] Unable to export object.
ERROR[/home/janisozaur/workspace/OpenRCT2/src/localisation/localisation.c:1443 (win1252_to_utf8)]: Encountered invalid sequence
<repeated numerous times>
[BRBASEV ] Error: Bad string table.
[BRBASEV ] Error: Unexpectedly reached end of file.
[BRBASEV ] Unable to export object.
[GWWTRHWF] Error: Bad string table.
[GWWTRHWF] Error: Unexpectedly reached end of file.
[GWWTRHWF] Unable to export object.
<repeated numerous times>
ERROR[/home/janisozaur/workspace/OpenRCT2/src/localisation/localisation.c:1443 (win1252_to_utf8)]: Encountered invalid sequence
<repeated numerous times>
[LVWATER1] Error: Bad image table.
[LVWATER1] Error: Unexpectedly reached end of file.
[LVWATER1] Unable to export object.
<repeated numerous times>
ERROR[/home/janisozaur/workspace/OpenRCT2/src/localisation/localisation.c:1443 (win1252_to_utf8)]: Encountered invalid sequence
<repeated numerous times>
[TTFALLSC] Error: Bad string table.
[TTFALLSC] Error: Unexpectedly reached end of file.
[TTFALLSC] Unable to export object.
[XXRQST11] Error: Bad string table.
[XXRQST11] Error: Unexpectedly reached end of file.
[XXRQST11] Unable to export object.
[XXRQST12] Error: Bad string table.
[XXRQST12] Error: Unexpectedly reached end of file.
[XXRQST12] Unable to export object.
[AE-DINVT] Object not found.
[AE-LOOP2] Object not found.
[HOPPER01] Object not found.
<repeated numerous times>
```

**Comment by Goddesen** [comment]
`if (!ret)` derp... I'll try to look more into this, but I've had real world stuff come up again 👎

**Comment by janisozaur** [comment]
Closing it in favour of https://github.com/OpenRCT2/OpenRCT2/pull/4938.

This can still get revived and merged at some point in future, but it isn't working now and we're trying to cut down on number of PRs.

Please see https://github.com/OpenRCT2/OpenRCT2/pull/4938#issuecomment-269506495

**Pull request closed without merging**